cdc/sink/manager.go

```diff
@@ -0,0 +1,267 @@
+	return atomic.LoadUint64(&m.checkpointTs)
+}
+
+type tableSink struct {
+	tableID model.TableID
+	manager *Manager
+	buffer  []*model.RowChangedEvent
+	// emittedTs means all of events which of commitTs less than or equal to emittedTs is sent to backendSink
+	emittedTs model.Ts
+}
+
+func (t *tableSink) Initialize(ctx context.Context, tableInfo []*model.SimpleTableInfo) error {
+	// do nothing
+	return nil
+}
+
+func (t *tableSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	t.buffer = append(t.buffer, rows...)
+	return nil
+}
+
+func (t *tableSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	// the table sink doesn't receive the DDL event
+	return nil
+}
+
+func (t *tableSink) FlushRowChangedEvents(ctx context.Context, resolvedTs uint64) (uint64, error) {
+	i := sort.Search(len(t.buffer), func(i int) bool {
+		return t.buffer[i].CommitTs > resolvedTs
+	})
+	if i == 0 {
+		atomic.StoreUint64(&t.emittedTs, resolvedTs)
+		return t.manager.flushBackendSink(ctx)
+	}
+	resolvedRows := t.buffer[:i]
+	t.buffer = t.buffer[i:]
+
+	err := t.manager.backendSink.EmitRowChangedEvents(ctx, resolvedRows...)
+	if err != nil {
+		return t.manager.getCheckpointTs(), errors.Trace(err)
+	}
+	atomic.StoreUint64(&t.emittedTs, resolvedTs)
+	return t.manager.flushBackendSink(ctx)
+}
+
+func (t *tableSink) getEmittedTs() uint64 {
+	return atomic.LoadUint64(&t.emittedTs)
+}
+
+func (t *tableSink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
+	// the table sink doesn't receive the checkpoint event
+	return nil
+}
+
+func (t *tableSink) Close() error {
+	t.manager.destroyTableSink(t.tableID)
+	return nil
+}
+
+type bufferSink struct {
+	Sink
+	buffer chan struct {
+		rows       []*model.RowChangedEvent
+		resolved   bool
+		resolvedTs model.Ts
+	}
+	checkpointTs uint64
+}
+
+func newBufferSink(ctx context.Context, backendSink Sink, errCh chan error, checkpointTs model.Ts) Sink {
+	sink := &bufferSink{
+		Sink: backendSink,
+		buffer: make(chan struct {
+			rows       []*model.RowChangedEvent
+			resolved   bool
+			resolvedTs model.Ts
+		}, defaultBufferChanSize),
+		checkpointTs: checkpointTs,
+	}
+	go sink.run(ctx, errCh)
+	return sink
+}
+
+func (b *bufferSink) run(ctx context.Context, errCh chan error) {
+	changefeedID := util.ChangefeedIDFromCtx(ctx)
+	advertiseAddr := util.CaptureAddrFromCtx(ctx)
+	metricFlushDuration := flushRowChangedDuration.WithLabelValues(advertiseAddr, changefeedID)
+	metricBufferSize := bufferChanSizeGauge.WithLabelValues(advertiseAddr, changefeedID)
+	for {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err != nil && errors.Cause(err) != context.Canceled {
+				errCh <- err
+			}
+			return
+		case e := <-b.buffer:
+			if e.resolved {
+				start := time.Now()
+				checkpointTs, err := b.Sink.FlushRowChangedEvents(ctx, e.resolvedTs)
+				if err != nil {
+					if errors.Cause(err) != context.Canceled {
+						errCh <- err
```

Comment:
`SinkManager` shares the same `errCh` with backend sink, in some extreme scenarios, both of them may try to send error to this channel, then one of them will hang